File: composeApp/src/androidMain/kotlin/io/github/tshion/trykmp/sample/MainViewModel.kt
package io.github.tshion.trykmp.sample

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.ViewModelStoreOwner
import androidx.lifecycle.viewModelScope
import androidx.lifecycle.viewmodel.CreationExtras
import androidx.lifecycle.viewmodel.MutableCreationExtras
import androidx.lifecycle.viewmodel.initializer
import androidx.lifecycle.viewmodel.viewModelFactory
import io.github.tshion.trykmp.TryKmp
import io.github.tshion.trykmp.timer
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class MainViewModel(
    private val model: TryKmp,
) : ViewModel() {

    private val _repos = MutableStateFlow("")
    val repos = _repos.asStateFlow()

    private val _timeText = MutableStateFlow("")
    val timeText = _timeText.asStateFlow()


    init {
        viewModelScope.launch {
            model.timer().collect {
                _timeText.value = it
            }
        }
    }


    fun searchRepo() {
        _repos.value = "LOADING..."

        val query = "android"
        viewModelScope.launch {
            val result = model.searchGitHubRepo(query)
            _repos.value = result
        }
    }


    companion object {
        private val KEY_TRYKMP = object : CreationExtras.Key<TryKmp> {}

        private val Factory = viewModelFactory {
            initializer {
                val tryKmp = this[KEY_TRYKMP] as TryKmp
                MainViewModel(tryKmp)
            }
        }


        fun create(
            tryKmp: TryKmp,
            viewModelStoreOwner: ViewModelStoreOwner,
        ) = ViewModelProvider.create(
            viewModelStoreOwner,
            factory = Factory,
            extras = MutableCreationExtras().apply {
                set(KEY_TRYKMP, tryKmp)
            }
        )[MainViewModel::class]
    }
}

File: iosApp/iosApp/MainViewModel.swift
import SwiftUI
import TryKMP

@MainActor
class MainViewModel: ObservableObject {

    private let model: TryKmp

    @Published private(set) var repos = ""

    private var timer: Timer? = nil

    @Published private(set) var timeText = ""


    init(_ model: TryKmp) {
        self.model = model
        timeText = model.time(suffix: "")
    }

    convenience init() {
        self.init(TryKmp())
    }


    func searchRepo() async {
        repos = "LOADING..."

        let query = "ios"
        let result = try! await model.searchGitHubRepo(query: query)
        repos = result
    }

    func startTimer() {
        stopTimer()
        timer = Timer.scheduledTimer(
            withTimeInterval: 2.5,
            repeats: true
        ) { _ in
            DispatchQueue.main.async {
                self.timeText = self.model.time(suffix: "")
            }
        }
    }

    func stopTimer() {
        timer?.invalidate()
        timer = nil
    }
}
